docs/madr/decisions/056-meshservice-migration.md

```diff
@@ -0,0 +1,349 @@
+# MeshService migration
+
+- Status: accepted
+
+## Context and Problem Statement
+
+This MADR addresses the switch from not having MeshServices at all to MeshServices
+being automatically created, either from `Services` in Kubernetes or from
+`Dataplanes` in Universal.
+
+When upgrading from 2.8 to 2.9, nothing at the data plane level should change.
+But MeshService behaves differently in several ways.
+
+In general the migration process for MeshService consists of no longer using
+`kuma.io/service` so:
+
+- targeting policies to the real `MeshService` resource
+- switching over to the `MeshService` i.e. `HostnameGenerator` hostnames and
+  local-zone load balancing from cross-zone load balancing
+
+Ideally users can try out `MeshService` behavior for individual consumers.
+
+Note that we in general can't be smart about which behavior should be the
+default because we can't differentiate between a new user, who doesn't care
+about `kuma.io/service`-style behavior, and an existing user who's just
+redeploying on a new `Mesh` or control plane.
+
+We have to make a choice between:
+
+- new behavior by default, `UPGRADE.md` tells users to set a variable to keep
+  old behavior
+- old behavior by default, new users have to be convinced to set a variable to
+  get new behavior
+
+### `MeshService`-targeted policy
+
+If for example we have a policy targeted at:
+
+```
+spec:
+  to:
+  - targetRef:
+      kind: MeshService
+      name: demo-app_kuma-demo_svc_5000
+```
+
+when switching over to MeshService, this policy will no longer apply.
+
+We can always convert the above to an additional entry:
+
+```
+  - targetRef:
+      kind: MeshService
+      labels:
+        kuma.io/display-name: demo-app
+        kuma.io/namespace: kuma-demo
+      port: 5000
+```
+
+Potentially we could offer a tool to convert these refs from old-style to
+new-style `MeshService`.
+
+#### Require `sectionName`
+
+Because the original refers to instances of `demo-app_kuma-demo_svc_5000` from
+any zone, we have to use `labels` to match instances outside of the local
+zone.
+
+Note that there's an ambiguity here with universal. We can't tell whether:
+
+```
+spec:
+  to:
+  - targetRef:
+      kind: MeshService
+      name: backend
+```
+
+is meant to target `kuma.io/service` or a `MeshService` named `backend` and thus
+we can't determine whether all instances of `kuma.io/service: backend` are meant
+or only the instances matched by `MeshService`.
+
+Thus we have to require `sectionName` to a port, in order to refer to
+`MeshServices` unambiguously.
+
+#### Port name
+
+We can't require users to add a `name` field to their existing `Services` or
+annotate their `Dataplanes` for universal, so we should
+set a canonical name for every port assuming one isn't set.
+
+Because the `port` field must be unique anyway, this could be a suitable,
+automatic `name`.
```

Comment:
Do we add this explicitly when we convert Service -> MeshService and Dataplane -> MeshService?

Reply:
yeah